### ISSUE TITLE
[bug] Create lock file from another lock file with aliases involved does not keep the fixed version

### ISSUE THREAD

**Issue by fourbft** [conan-io/conan#9014]
### Environment Details (include every applicable attribute)
  * Operating System+version: Shouldn't matter, but seen on Linux
  * Compiler+version: No compiler involved
  * Conan version: 1.35.0 and 1.36.0
  * Python version: 3.6

### Steps to reproduce (Include if Applicable)

Basically what goes wrong is the following:
Create a package CA with version 1.0.0 (no contents, just the recipe with empty build and package methods.)
Create an alias to this package CA/LATEST-master@user/testing that points to that version.

Create a package recipe for a component CB with verision 1.0.0 that requires CA/LATEST-master@user/testing.
Again no further contents.
Create that package.

Create a conanfile.txt:
```
[requires]
CA/LATEST-master@user/testing
CB/LATEST-master@user/testing
```

Create a lockfile for this conanfile.txt.
Now create a new version of CA (2.0.0) and update the alias CA/LATEST-master@user/testing to point to the new version.
Then go to the recipe of CB, update the version in there to 1.0.1. Then create a new lockfile from the conanfile.txt lockfile. You will notice, that the version of CA locked in the new lockfile is 2.0.0 instead of keeping CA at version 1.0.0 as locked in the input lockfile.

When performing the same procedure with version ranges instead of aliases (`self.requires( "CA/[>=1.0.0]@user/testing")` in CB) it all works as expected, i. e. when creating the second lockfile, CA is kept at version 1.0.0.

The expectation is, that this should work the same with aliases.

To reproduce, you can unpack the attached ZIP and do a:
```
./create_segment_lockfile_and_use_it_for_component_build-version_ranges.sh
```
Which shows the good case with version ranges. Then:
```
./cleanup.sh
./create_segment_lockfile_and_use_it_for_component_build-aliases.sh
```
Which demonstrates that in the case of aliases, the version of CA is not locked correctly, but the latest version of CA is used.

**Comment by fourbft** [comment]
Sorry, cannot upload the zip file with the scripts to reproduce the error. My browser keeps telling me: "Something went really wrong, and we can't process that file. Try again". What am I doing wrong. Tried tar.gz an .zip files. Both don't seem to work...

**Comment by fourbft** [comment]
Trying with IE ...

**Comment by fourbft** [comment]
Sorry, doesn't work either. File selection dialog appears, but after selecting the file it is not accepted ...

**Comment by fourbft** [comment]
Ok, then I paste the code for the script to reproduce the bug here:
```
#!/bin/bash

# This test does the following:
# Build two components CA <- CB where CB uses CA.
# Create all components and the LATEST-master aliases through
# which CB uses CA.
# Then create a segment-lock-file from a conanfile.txt which
# includes both packages. This lockfile locks the initial
# versions (1.0.0) of CA and CB.
# Then create a new version 2.0.0 of CA and move the LATEST-master
# alias to the new version.
#
# Then update CB from version 1.0.0 to 1.0.1 and create a new
# lock-file for CB from the segment-lock-file.
# Use this lock file to create CB 1.0.1 with the old version
# 1.0.0 of CA.
#
# Conan versions 1.35.0 and 1.36.0 incorrectly lock CA with version
# 2.0.0 and build CB-1.0.1 with the new CA version.
# When using version ranges instead of aliases this exact
# same procedure works. (Use create_segment_lockfile_and_use_it_for_component_build-version_ranges.sh
# that comes alongside this script to verify this.)
# Use the script cleanup.sh to clean up between the tests.

mkdir workdir
cd workdir

echo "#########################################################"
echo "Creating first version 1.0.0 of package CA"
echo "#########################################################"
mkdir CA
cd CA
cat > conanfile.py << EOF
from conans import ConanFile

class SpecificConanfile(ConanFile):
    name = "CA"
    description = "CA test component for lockfile test."
    url = "http://CA.html"
    version = "1.0.0"

    def requirements( self):
        pass

    def build( self):
        pass
EOF
conan create . user/testing
conan alias CA/LATEST-master@user/testing CA/1.0.0@user/testing

cd ..
echo "#########################################################"
echo "Creating first version of package CB, using CA 1.0.0"
echo "#########################################################"
mkdir CB
cd CB
cat > conanfile.py << EOF
from conans import ConanFile

class SpecificConanfile(ConanFile):
    name = "CB"
    description = "CB test component for lockfile test."
    url = "http://CB.html"
    version = "1.0.0"

    def requirements( self):
        self.requires( "CA/LATEST-master@user/testing")

    def build( self):
        pass
EOF
conan create . user/testing
conan alias CB/LATEST-master@user/testing CB/1.0.0@user/testing

cd ..
echo "#########################################################"
echo "Creating segment conanfile.txt with LATEST-master versions of CA and CB."
echo "#########################################################"
mkdir segment
cd segment
cat > conanfile.txt << EOF
[requires]
CA/LATEST-master@user/testing
CB/LATEST-master@user/testing
EOF

conan lock create conanfile.txt --lockfile-out lock/segment.lock

echo "#########################################################"
echo "Creating new version 2.0.0 of package CA"
echo "#########################################################"
cd ../CA
cat > conanfile.py << EOF
from conans import ConanFile

class SpecificConanfile(ConanFile):
    name = "CA"
    description = "CA test component for self.build_requires bug."
    url = "http://CA.html"
    version = "2.0.0"

    def requirements( self):
        pass

    def build( self):
        pass
EOF
conan create . user/testing
conan alias CA/LATEST-master@user/testing CA/2.0.0@user/testing

echo "#########################################################"
echo "Now build some changes of package CB, using the segment lockfile."
echo "#########################################################"
cd ../CB
cat > conanfile.py << EOF
from conans import ConanFile

class SpecificConanfile(ConanFile):
    name = "CB"
    description = "CB test component for self.build_requires bug."
    url = "http://CB.html"
    version = "1.0.1"

    def requirements( self):
        self.requires( "CA/LATEST-master@user/testing")

    def build( self):
        pass
EOF
echo "###### Creating lockfile for CB from segment lockfile."
conan lock create conanfile.py --user=user --channel=testing --lockfile=../segment/lock/segment.lock --lockfile-out=lock/CB-default_profile.lock
echo "###### Building with concrete lockfile."
conan create . user/testing --lockfile=lock/CB-default_profile.lock
```

**Comment by fourbft** [comment]
Working example with version ranges for comparison:
```
#!/bin/bash

mkdir workdir
cd workdir

# This test does the following:
# Build two components CA <- CB where CB uses CA.
# Create both components. CB requires CA using version
# range [>=1.0.0].
# Then create a segment-lock-file that locks the initial
# versions.
# Then create a new version (2.0.0) of CA.
#
# Then update CB from version 1.0.0 to 1.0.1 and create a new
# lock-file for CB from the segment-lock-file.
# Use this lock file to create CB 1.0.1 with the old version
# 1.0.0 of CA.
#
# With version ranges, this works. With aliases it does not.
# Execute create_segment_lockfile_and_use_it_for_component_build-aliases.sh
# which comes alongside this script to verify this.
# Use the script cleanup.sh to clean up between the tests.

echo "#########################################################"
echo "Creating first version 1.0.0 of package CA"
echo "#########################################################"
mkdir CA
cd CA
cat > conanfile.py << EOF
from conans import ConanFile

class SpecificConanfile(ConanFile):
    name = "CA"
    description = "CA test component for lockfile test."
    url = "http://CA.html"
    version = "1.0.0"

    def requirements( self):
        pass

    def build( self):
        pass
EOF
conan create . user/testing

cd ..
echo "#########################################################"
echo "Creating first version of package CB, using CA 1.0.0"
echo "#########################################################"
mkdir CB
cd CB
cat > conanfile.py << EOF
from conans import ConanFile

class SpecificConanfile(ConanFile):
    name = "CB"
    description = "CB test component for lockfile test."
    url = "http://CB.html"
    version = "1.0.0"

    def requirements( self):
        self.requires( "CA/[>=1.0.0]@user/testing")

    def build( self):
        pass
EOF
conan create . user/testing

cd ..
echo "#########################################################"
echo "Locking CB and CA with current versions."
echo "#########################################################"
mkdir segment
cd segment
cat > conanfile.txt << EOF
[requires]
CA/[>=1.0.0]@user/testing
CB/[>=1.0.0]@user/testing
EOF

conan lock create conanfile.txt --lockfile-out lock/segment.lock

echo "#########################################################"
echo "Creating new version 2.0.0 of package CA"
echo "#########################################################"
cd ../CA
cat > conanfile.py << EOF
from conans import ConanFile

class SpecificConanfile(ConanFile):
    name = "CA"
    description = "CA test component for self.build_requires bug."
    url = "http://CA.html"
    version = "2.0.0"

    def requirements( self):
        pass

    def build( self):
        pass
EOF
conan create . user/testing

echo "#########################################################"
echo "Now build some changes of package CB, using using the segment lockfile."
echo "#########################################################"
cd ../CB
cat > conanfile.py << EOF
from conans import ConanFile

class SpecificConanfile(ConanFile):
    name = "CB"
    description = "CB test component for self.build_requires bug."
    url = "http://CB.html"
    version = "1.0.1"

    def requirements( self):
        self.requires( "CA/[>=1.0.0]@user/testing")

    def build( self):
        pass
EOF
echo "###### Creating lockfile for CB from segment lockfile."
conan lock create conanfile.py --user=user --channel=testing --lockfile=../segment/lock/segment.lock --lockfile-out=lock/CB-default_profile.lock
echo "###### Building with concrete lockfile."
conan create . user/testing --lockfile=lock/CB-default_profile.lock
```

**Comment by memsharded** [comment]
Hi @fourbft 

Thanks for your reproducible case. I have translated it to a test, you might find this useful too in order to reproduce things:

```python
def test_lockfile_alias():
    # https://github.com/conan-io/conan/issues/9014
    client = TestClient()
    client.save({"conanfile.py": GenConanfile()})
    client.run("create . ca/1.0@user/testing")
    client.run("alias ca/latest@user/testing ca/1.0@user/testing")

    client.save({"conanfile.py": GenConanfile().with_requires("ca/latest@user/testing")})
    client.run("create . cb/1.0@user/testing")
    client.run("alias cb/latest@user/testing cb/1.0@user/testing")

    client.save({"conanfile.py": GenConanfile().with_requires("ca/latest@user/testing",
                                                              "cb/latest@user/testing")})

    client.run("lock create conanfile.py --lockfile-out=segment.lock")

    # Change CA latest
    client.save({"conanfile.py": GenConanfile()})
    client.run("create . ca/2.0@user/testing")
    client.run("alias ca/latest@user/testing ca/2.0@user/testing")

    # 1st CASE: When using same version, it can find it in the lockfile, and it is able to lock dependencies
    client.save({"conanfile.py": GenConanfile("cb", "1.0").with_requires("ca/latest@user/testing")})
    client.run("lock create conanfile.py --user=user --channel=testing "
               "--lockfile=segment.lock --lockfile-out=cbnew.lock")
    cbnew = client.load("cbnew.lock")
    assert "ca/1.0" in cbnew

    # 2nd CASE: When using new version, it cannot find it
    client.save({"conanfile.py": GenConanfile("cb", "2.0").with_requires("ca/latest@user/testing")})
    client.run("lock create conanfile.py --user=user --channel=testing "
               "--lockfile=segment.lock --lockfile-out=cbnew.lock")
    cbnew = client.load("cbnew.lock")
    assert "ca/2.0" in cbnew
```

It seems that this might be expected behavior:

First, you capture a lockfile from the first state. The resulting lockfile contains:
```
"1": {
    "ref": "ca/1.0@user/testing",
    ...
   },
   "2": {
    "ref": "cb/1.0@user/testing",
    ...
    "requires": [
     "1"
    ],
```

That is, it is resolving and storing the real versions in the graph.

As described in the docs, an alias is: 

> ``An “alias” package is a symbolic name (reference) for another package (target). ``


In the test first case, when you keep ``cb`` version to ``cb/1.0``,  Conan is able to locate that package in the previous lockfile, and obtain its dependency to ``ca/1.0``. Everything works.

But when the ``cb`` version is bumped, ``cb/2.0`` it is not located in the lockfile, as it is a new element. Things have changed, and ``cb/2.0`` is able to introduce new dependencies, remove dependencies, or change dependencies, and when it is evaluated, that should be respected. A lockfile is a snapshot in time of a state, but that state must be possible. What it is not possible is that ``cb/2.0`` for example had a dependency to ``ca/3.0`` and that would be overwritten by the lockfile to something that is impossible in the first place (``cb/2.0`` never in time had a dependency to ``ca/1.0``, it is impossible state). So a new snapshot around the new ``cb/2.0`` is computed. As it has an alias, the current alias will be resolved.

In summary it is not possible to have the 2 opposites behaviors: to allow changes and evolution from cb/1.0 => cb/2.0, while keeping a complete lock over cb/1.0 dependencies that is extrapolated to cb/2.0, even if it is incompatible.

Please let me know if this helps to clarify what is happening. Thanks!

**Comment by fourbft** [comment]
Hi @memsharded, thanks a lot for looking into this issue and thanks for the translation to a test. That will certainly come in handy in the future.
As for your explanation of the behavior: yes, it sounds logical and maybe that is the intended behavior from the current conan lockfile mechanism point of view. But if it is so, that when a new version of cb (2.0) is being used with the lockfile for cb/1.0 the locks do no longer apply, how does that logic match with the behavior in case of the version ranges?
As my version range example shows, when I do the same thing with version ranges and I bump the cb version from 1.0 to 2.0, the version of ca is still locked to 1.0 even though a ca/2.0 is available and is compatible with cb's requirements. Should conan not reevaluate these requirements then also and use ca/2.0?

Anyway, be it as it may, then we should probably change this issue from "bug" to "feature request". What we really need is a means to tell conan that we want upstreams of a package P (which are referenced via aliases in our case, but version ranges should work the same) to be of specific, older versions, no matter how the aliases are currently set on the conan server. And this should be independent of if P's version itself has changed or not.

Until conan versions before 1.29 this use case was perfectly covered by conan's lockfile mechanism. From 1.29 on, this is obviously no longer the case. So would it be possible to reenable this use case - either by special command line options to the lockfile mechanism (e. g. something like --ignore-version-change-of P) or by a new mechanism that resembles the old lockfile behavior?

**Comment by memsharded** [comment]
Closing this ticket as outdated, Conan 2.0 is now 1 year old, and its lockfiles are way simpler and more powerful, please create new tickets as necessary regarding the new 2.0 lockfiles, thanks!
